x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/action/PutEnrichPolicyAction.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.enrich.action;
+
+import org.elasticsearch.action.Action;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.action.support.master.MasterNodeRequest;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.enrich.EnrichPolicy;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class PutEnrichPolicyAction extends Action<AcknowledgedResponse> {
+    public static final PutEnrichPolicyAction INSTANCE = new PutEnrichPolicyAction();
+    public static final String NAME = "cluster:admin/xpack/enrich/put";
+
+    protected PutEnrichPolicyAction() {
+        super(NAME);
+    }
+
+    public static Request fromXContent(XContentParser parser, String name) throws IOException {
+        return new Request(name, EnrichPolicy.fromXContent(parser));
+    }
+
+    @Override
+    public AcknowledgedResponse newResponse() {
+        return new AcknowledgedResponse();
+    }
+
+    public static class Request extends MasterNodeRequest<PutEnrichPolicyAction.Request> {
+        private final EnrichPolicy policy;
+        private final String name;
+
+
+        public Request(String name, EnrichPolicy policy) {
+            this.name = Objects.requireNonNull(name, "name cannot be null");
+            this.policy = policy;
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            name = in.readString();
+            policy = new EnrichPolicy(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(name);
+            policy.writeTo(out);
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public EnrichPolicy getPolicy() {
+            return policy;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
```

Comment:
add validation here that the name is required?

Reply:
im not sure we need this since the constructor for the request has an `Objects.requireNotNull`. I could add something to the streaminput constructor as well, but that does not look like its typically done, as we should be able to trust that more than a "user input" constructor, so to speak.

Reply:
Good point. Maybe instead of using `Objects.requireNotNull`, use an if statement that checks now null and if so then throw a `IllegalArgumentException`? `Objects.requireNotNull` throws a NPE and that will result in a 500 response code, whilst a `IllegalArgumentException` translates to a 400 response code and that is a better response code in this case.

Reply:
Never mind. This good as it is now.

In production, the constructor can't be invoked with a null `name`, because if the rest handler is invoked then the  `name` had to provided.